# Introduction to "Exploring Ebay Car Sales Data"
### Project using NumPy and Pandas libraries with Jupyter notebook for data cleaning and analysis

In this project we will be looking into a dataset of 50k datapoints on used cars ads scraped from Ebay Kleinanzeigen, Germany.

The dataset columns are ordered in the following way:

**dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
**name** - Name of the car.
**seller** - Whether the seller is private or a dealer.
**offerType*+ - The type of listing
**price** - The price on the ad to sell the car.
**abtest** - Whether the listing is included in an A/B test.
**vehicleType** - The vehicle Type.
**yearOfRegistration** - The year in which which year the car was first registered.
**gearbox** - The transmission type.
**powerPS** - The power of the car in PS.
**model** - The car model name.
**kilometer** - How many kilometers the car has driven.
**monthOfRegistration** - The month in which which year the car was first registered.
**fuelType** - What type of fuel the car uses.
**brand** - The brand of the car.
**notRepairedDamage** - If the car has a damage which is not yet repaired.
**dateCreated** - The date on which the eBay listing was created.
**nrOfPictures** - The number of pictures in the ad.
**postalCode** - The postal code for the location of the vehicle.
**lastSeenOnline** - When the crawler saw this ad last online.

In [769]:
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv",encoding="Latin-1")
autos["price"] = autos["price"].str.replace("$","€")

In [770]:
autos.info()
autos.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  €5,000  control         bus                2004   
1  privat   Angebot  €8,500  control   limousine                1997   
2  privat   Angebot  €8,990     test   limousine                2009   
3  privat   Angebot  €4,350  control  kleinwagen                2007   
4  privat   Angebot  €1,350     test       kombi                2003   
5  privat   Angebot  €7,900     test         bus                2006   
6  privat   Angebot    €300     test   limousine                1995   
7  privat   Angebot  €1,990  control   limousine                1998   
8  privat   Angebot    €250     test         NaN                2000   
9  privat   Angebot    €590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

### First observations:
- It seems like Ebay Kleinanzeigen is a preferred platform for privates to put up their ads without using an agency or a car stand to promote their sale. We will look into this with more detailed in the upcoming cells.
- Price tags were replaced with € instead of $.
- We looked into the number of null values per columns and, in the worst case (notRepairedDamage), they don't account for more than 20% of the column's total.

### CamelCase to snake_case
Columns' names are not using Python's preferred snakecase naming system so we'll look into it in the next cells.

In [771]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [772]:
new_cols = ["date_crawled", "name", "seller", "offer_type", "price", "ab_test", "vehicle_type", "registration_year", "gear_box", "power_PS", "model", "odometer", "registration_month", "fuel_type", "brand", "unrepaired_damage", "ad_created", "n_pictures", "postal_code", "last_seen"]

As the encoding of this dataset is not UTF-8 we had to find another way around to rename all the columns since the usual df = df[lst] produced an error in Python:

In [773]:
autos.columns = new_cols

In [774]:
autos.head(10)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  €5,000  control          bus               2004   
1  privat    Angebot  €8,500  control    limousine               1997   
2  privat    Angebot  €8,990     test    limousine               2009   
3  privat    Angebot  €4,350  control   kleinwagen               2007   
4  privat    Angebot  €1,350     test        kombi               2003   
5  privat    Angebot  €7,900     test          bus               2006   
6  privat    Angebot    €300     test    limousine               1995   
7  privat    Angebot  €1,990  control    limousine               1998   
8  privat    Angebot    €250     test          NaN               2000   
9  privat    Angebot    €590  control          bus               1997   

    gear_box  power_PS    model   odometer  registration_month fuel_type  \
0    manuell       158   andere  150,000km                   3       lpg   
1  automatik       286      7er  150,000km                   6    benzin   
2    manuell       102     golf   70,000km                   7    benzin   
3  automatik        71   fortwo   70,000km                   6    benzin   
4    manuell         0    focus  150,000km                   7    benzin   
5  automatik       150  voyager  150,000km                   4    diesel   
6    manuell        90     golf  150,000km                   8    benzin   
7    manuell        90     golf  150,000km                  12    diesel   
8    manuell         0    arosa  150,000km                  10       NaN   
9    manuell        90   megane  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  n_pictures  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   
5    chrysler               NaN  2016-03-21 00:00:00           0        22962   
6  volkswagen               NaN  2016-03-20 00:00:00           0        31535   
7  volkswagen              nein  2016-03-16 00:00:00           0        53474   
8        seat              nein  2016-03-22 00:00:00           0         7426   
9     renault              nein  2016-03-16 00:00:00           0        15749   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50  
5  2016-04-06 09:45:21  
6  2016-03-23 02:48:59  
7  2016-04-07 03:17:32  
8  2016-03-26 18:18:10  
9  2016-04-06 10:46:35

Now that all columns have been renamed for easier indexing and identification, we can proceed with some basic data exploration to understand how much more cleaning this dataset would be before analysis.

### Data exploration and cleaning

Let's start with some basic clean up, translation and columns drop.

We'll also get a general overview on the descriptive statistics of the dataset using the df.describe() method for all columns (object and int/float).

In [775]:
autos = (autos
            .replace("ja","yes")
            .replace("nein","no")
            .replace("privat","private")
            .replace("manuell","manual")
            .replace("automatik","automatic")
            .replace("andere","other")
            .replace("benzin","gasoline")
            .replace("Angebot","deal")
        )

In [776]:
autos = autos.drop(columns=["offer_type","ab_test","seller","n_pictures"])

In [777]:
autos.describe(include="all")

date_crawled         name  price vehicle_type  \
count                 50000        50000  50000        44905   
unique                48213        38754   2357            8   
top     2016-03-12 16:06:22  Ford_Fiesta     €0    limousine   
freq                      3           78   1421        12859   
mean                    NaN          NaN    NaN          NaN   
std                     NaN          NaN    NaN          NaN   
min                     NaN          NaN    NaN          NaN   
25%                     NaN          NaN    NaN          NaN   
50%                     NaN          NaN    NaN          NaN   
75%                     NaN          NaN    NaN          NaN   
max                     NaN          NaN    NaN          NaN   

        registration_year gear_box      power_PS  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN   manual           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN  gasoline  volkswagen                no   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [778]:
autos["price"] = (autos["price"]
                      .str.replace("€","")
                      .str.replace(",","")
                      .str.strip()
                      .astype(int)
                 )

autos.rename(columns={"price": "price_€"}, inplace=True)

autos["odometer"] = (autos["odometer"]
                        .str.replace("km","")
                        .str.replace(",","")
                        .str.strip()
                        .astype(int)
                    )

autos.rename(columns={"odometer":"odometer_km"}, inplace=True)

In the first steps of this data exploration it was found that some of the columns seem to be irrelevant as they have only 1 real value, eg: in "seller" there is only "privat" as value; "offer_type" only produces "Angebot".

Also, since **price** and **odometer** columns still had a type "object" (string) in the dataframe, these were cleaned up and updated to type int, in order to better work with the values.
As these are two of the most interesting columns to analyze, let's start by doing some prep work for the "price_€" and "odometer_km" columns:

In [779]:
autos["price_€"].unique().shape

(2357,)

In [780]:
autos["price_€"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_€, dtype: float64

In [781]:
autos["price_€"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_€, dtype: int64

For the 2357 unique values in the "price_€" column, it was noticed that some of the highest and lowest values seemed unrealistic, so we'll treat them as outliers.

In [782]:
autos = autos[autos["price_€"] > 500]
autos = autos[autos["price_€"] < 350000]

In [783]:
autos["price_€"].unique().shape

(2206,)

In [784]:
autos["price_€"].value_counts().sort_index(ascending=False).head(20)

345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
194000    1
190000    1
180000    1
175000    1
169999    1
169000    1
163991    1
163500    1
155000    1
151990    1
145000    1
Name: price_€, dtype: int64

We now saw a decrease of 6% of the total dataset, after cleaning up the outliers from the **price** column. This was decided after seeing an increment of almost more than 3 fold from the 350k value to the following one. Realistically speaking, there shouldn't be any cars in that price range, at least for this kind of market, and specially in this ad platform.       
Let's now look a bit deeper into the **odometer** column to see if we can find some outliers there.

In [785]:
autos["odometer_km"].describe()

count     44315.000000
mean     125052.239648
std       39720.642964
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [786]:
autos["odometer_km"].unique().shape

(13,)

In [787]:
autos["odometer_km"].value_counts().sort_index(ascending=False).head(13)

150000    28056
125000     4770
100000     2014
90000      1665
80000      1381
70000      1187
60000      1126
50000       994
40000       806
30000       764
20000       724
10000       236
5000        592
Name: odometer_km, dtype: int64

Even though a used car with 5000km is not that common it wouldn't also be considered unrealistic. Besides, there seems to be a healthy increment between the 13 unique values. For this reason, we'll leave this column as it is for now and proceed.

In [788]:
autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(20)

2016-03-05 14:06:30    0.000023
2016-03-05 14:06:40    0.000023
2016-03-05 14:07:08    0.000023
2016-03-05 14:07:21    0.000023
2016-03-05 14:07:26    0.000023
2016-03-05 14:07:40    0.000023
2016-03-05 14:07:45    0.000023
2016-03-05 14:08:00    0.000045
2016-03-05 14:08:05    0.000045
2016-03-05 14:08:42    0.000023
2016-03-05 14:09:02    0.000045
2016-03-05 14:09:05    0.000023
2016-03-05 14:09:22    0.000023
2016-03-05 14:09:38    0.000023
2016-03-05 14:09:46    0.000023
2016-03-05 14:09:57    0.000023
2016-03-05 14:09:58    0.000045
2016-03-05 14:10:18    0.000023
2016-03-05 14:10:20    0.000023
2016-03-05 14:10:46    0.000023
Name: date_crawled, dtype: float64

In [789]:
autos["ad_created"] = autos["ad_created"].str.replace("-","")
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(20)

20150611 00:00:00    0.000023
20150810 00:00:00    0.000023
20150909 00:00:00    0.000023
20151110 00:00:00    0.000023
20151205 00:00:00    0.000023
20151230 00:00:00    0.000023
20160103 00:00:00    0.000023
20160107 00:00:00    0.000023
20160110 00:00:00    0.000045
20160113 00:00:00    0.000023
20160114 00:00:00    0.000023
20160116 00:00:00    0.000023
20160122 00:00:00    0.000023
20160127 00:00:00    0.000068
20160129 00:00:00    0.000023
20160201 00:00:00    0.000023
20160202 00:00:00    0.000045
20160205 00:00:00    0.000045
20160207 00:00:00    0.000023
20160208 00:00:00    0.000023
Name: ad_created, dtype: float64

In [790]:
autos["last_seen"] = autos["last_seen"].str.replace("-","")
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index(ascending=True).head(20)

20160305 14:45:46    0.000023
20160305 14:46:02    0.000023
20160305 14:49:34    0.000023
20160305 15:16:11    0.000023
20160305 15:16:47    0.000023
20160305 15:28:10    0.000023
20160305 15:41:30    0.000023
20160305 15:45:43    0.000023
20160305 15:47:38    0.000023
20160305 15:47:44    0.000023
20160305 16:45:57    0.000023
20160305 16:47:28    0.000023
20160305 17:15:45    0.000023
20160305 17:16:14    0.000023
20160305 17:16:23    0.000023
20160305 17:17:02    0.000023
20160305 17:39:19    0.000023
20160305 17:40:14    0.000023
20160305 17:44:50    0.000023
20160305 17:44:54    0.000023
Name: last_seen, dtype: float64

In [791]:
autos["last_seen"] = autos["last_seen"].apply(lambda x: x.split()[0])
autos["ad_created"] = autos["ad_created"].apply(lambda x: x.split()[0])
autos["date_crawled"] = autos["date_crawled"].apply(lambda x: x.split()[0]).str.replace("-","")

In [792]:
autos["registration_year"].describe()

count    44315.000000
mean      2005.112355
std         89.450654
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [793]:
autos = autos[autos["registration_year"] > 1900]
autos = autos[autos["registration_year"] < 2017]

Dates in the last_seen, ad_created and date_crawled were still represented as strings. As an initial approach to the dates clean up the "-" character has been replaced. This should make it easier to analyze afterwards.

Also, from an initial look into the registration year values, it was decided we should keep the years between 1900 and 2017.

In [794]:
autos["registration_year"].value_counts(normalize=True).head(20)

2005    0.067144
2004    0.062869
2006    0.062610
2003    0.061999
2000    0.061882
2001    0.058264
1999    0.058123
2002    0.056243
2007    0.053283
2008    0.051827
2009    0.048819
1998    0.045812
2011    0.037989
2010    0.037261
1997    0.033102
2012    0.030682
1996    0.023564
2016    0.021356
2013    0.018724
1995    0.018513
Name: registration_year, dtype: float64

In [795]:
autos["registration_year"].describe()

count    42565.000000
mean      2003.316880
std          7.051594
min       1910.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

The describe method tells us that most of the advertised cars are newer than 2003. Also, it seems there is a vintage car, from 1910! Let's check it out:

In [796]:
autos.loc[autos["registration_year"] == 1910] 

date_crawled            name  price_€ vehicle_type  registration_year  \
28693     20160322  Renault_Twingo      599   kleinwagen               1910   

      gear_box  power_PS model  odometer_km  registration_month fuel_type  \
28693   manual         0   NaN         5000                   0  gasoline   

         brand unrepaired_damage ad_created  postal_code last_seen  
28693  renault               NaN   20160322        70376  20160406

Oh, it's clearly a false value as Twingos are not older than 1992. Let's update this to the mean value.

In [797]:
autos["registration_year"].replace(1910, np.mean, inplace=True)

Ok, the house is now a little bit tidier.

As we were speaking of Twingos, let us now delve deeper into brands and their mean prices.

### Brand Analysis

Firstly, let's check which brands are worth analyzing. For this we'll look into the uniques as well as the ones with higher frequency (>= 5%) in our dataset.

In [798]:
autos["brand"].value_counts(normalize=True).sort_index()

alfa_romeo        0.006461
audi              0.091108
bmw               0.116175
chevrolet         0.006108
chrysler          0.003454
citroen           0.014214
dacia             0.002890
daewoo            0.001339
daihatsu          0.002161
fiat              0.023329
ford              0.064349
honda             0.007706
hyundai           0.010196
jaguar            0.001645
jeep              0.002490
kia               0.007306
lada              0.000611
lancia            0.000963
land_rover        0.002302
mazda             0.014777
mercedes_benz     0.103418
mini              0.009538
mitsubishi        0.007612
nissan            0.014801
opel              0.099706
peugeot           0.029226
porsche           0.006508
renault           0.043158
rover             0.001222
saab              0.001645
seat              0.017926
skoda             0.017315
smart             0.015318
sonstige_autos    0.009750
subaru            0.001926
suzuki            0.005967
toyota            0.013509
t

In [799]:
selected_brands = ["audi", "bmw", "citroen", "ford", "mercedes_benz", "opel", "volkswagen"]

In [800]:
brand_mean_price = {}

for each in selected_brands:
    selected_rows = autos[autos["brand"] == each].sort_values("brand",ascending=False)
    mean = selected_rows["price_€"].mean().astype(int)
    brand_mean_price[each] = mean

In [801]:
brand_mean_price

{'audi': 9714,
 'bmw': 8644,
 'citroen': 4059,
 'ford': 4407,
 'mercedes_benz': 8819,
 'opel': 3464,
 'volkswagen': 5886}

From this analysis we extracted the mean prices of each of the selected brands and found out that the most expensive brands are, in descending order, **Audi**, **Mercedes Benz**, and **BMW**; while **Ford** and **Opel** are the less expensive.

But a question arises: are these brands more expensive because they are "better" or because those specific listed cars are simply newer? Let's compare the mean prices with the mean mileage so we can better understand this.

In [802]:
brand_mean_mileage = {}

for each in selected_brands:
    selected_rows = autos[autos["brand"] == each].sort_values("brand",ascending=False)
    mean = selected_rows["odometer_km"].mean().astype(int)
    brand_mean_mileage[each] = mean

In [803]:
brand_mean_mileage

{'audi': 128713,
 'bmw': 132734,
 'citroen': 118165,
 'ford': 123121,
 'mercedes_benz': 130970,
 'opel': 127725,
 'volkswagen': 127948}

In [804]:
bmp_series = pd.Series(brand_mean_price)

In [805]:
bmm_series = pd.Series(brand_mean_mileage)

In [806]:
data_dict = {"mean_price": bmp_series, "mean_mileage_km": bmm_series, "ratio price per km": bmp_series / bmm_series} 
df_means = pd.DataFrame(data_dict) 
df_means.sort_values("ratio price per km", ascending=False)

mean_price  mean_mileage_km  ratio price per km
audi                 9714           128713            0.075470
mercedes_benz        8819           130970            0.067336
bmw                  8644           132734            0.065123
volkswagen           5886           127948            0.046003
ford                 4407           123121            0.035794
citroen              4059           118165            0.034350
opel                 3464           127725            0.027121

So, it appears that not only the price is higher for the aforementioned top 3 brands but also their mean mileage per km. A ratio of price per km was calculated as well and sorted in descendent order.

In [812]:
import matplotlib.pyplot as plt

In [ ]:
fig = plt.figure(figsize=(12,12))

plt.plot(df_means[])

### Most common brand + model combination

In [807]:
autos["top_brandmodel"] = autos["brand"].map(str) + "_" + autos["model"].map(str)
  

In [808]:
autos["top_brandmodel"].value_counts().head()

volkswagen_golf      3361
bmw_3er              2490
volkswagen_polo      1300
volkswagen_passat    1291
opel_corsa           1221
Name: top_brandmodel, dtype: int64

VW has two of their models on the 1st and 3rd positions of the most common advertised cars, with the BMW Series 3 coming in second place.

### Prices of damaged cars vs not damaged

In [809]:
autos["unrepaired_damage"].describe()

count     35997
unique        2
top          no
freq      32691
Name: unrepaired_damage, dtype: object

In [810]:
autos["damaged_cars"] = autos["unrepaired_damage"].where(autos["unrepaired_damage"] == "yes")
autos["non_damaged_cars"] = autos["unrepaired_damage"].where(autos["unrepaired_damage"] == "no")

autos.loc[autos["damaged_cars"] == "yes"].head()

date_crawled                                  name  price_€ vehicle_type  \
28      20160319                         MINI_Cooper_D     5250   kleinwagen   
51      20160322               Mercedes_Benz_S_320_CDI     6000    limousine   
81      20160312  Nissan_Micra_K12___super_Kleinwagen!     2000   kleinwagen   
97      20160401            Treuer_Begleiter_abzugeben      800          bus   
100     20160312          Mazda_5_2.0_CD_DPF_Exclusive     4500          bus   

    registration_year   gear_box  power_PS     model  odometer_km  \
28               2007     manual       110    cooper       150000   
51               2005  automatic       204  s_klasse       150000   
81               2007     manual        65     micra       150000   
97               2000     manual        63     other       125000   
100              2007     manual       143   5_reihe        90000   

     registration_month fuel_type          brand unrepaired_damage ad_created  \
28                    7    diesel           mini               yes   20160319   
51                    7    diesel  mercedes_benz               yes   20160322   
81                    9  gasoline         nissan               yes   20160312   
97                    8  gasoline          mazda               yes   20160401   
100                   7    diesel          mazda               yes   20160312   

     postal_code last_seen          top_brandmodel damaged_cars  \
28         15745  20160407             mini_cooper          yes   
51         49492  20160404  mercedes_benz_s_klasse          yes   
81         52156  20160326            nissan_micra          yes   
97         87648  20160405             mazda_other          yes   
100        51375  20160407           mazda_5_reihe          yes   

    non_damaged_cars  
28               NaN  
51               NaN  
81               NaN  
97               NaN  
100              NaN

In [811]:
damage = {}
damage_nodamage = ["yes","no"]

for each in damage_nodamage:
    selected_rows = autos[autos["unrepaired_damage"] == each]
    price = selected_rows["price_€"]
    damage[each] = price
    
damage_df = pd.DataFrame(damage)
damage_df.describe()

yes             no
count   3306.000000   32691.000000
mean    2957.128252    7391.509162
std     3942.554234    9992.911014
min      520.000000     501.000000
25%      900.000000    1999.000000
50%     1500.000000    4400.000000
75%     3499.000000    9450.000000
max    44200.000000  345000.000000

Cars without with some unrepaired damages are in average 40% lower in price.